README.md
![IMDB](http://www.digifail.com/images/misc/github/imdb_logo_sm.jpg "IMDB Logo")

imdb_scrape
==============

imdb_scrape is a simple Bash script to "scrape" the user-submitted data on the
Internet Movie Database (www.imdb.com), namely, the "Parents Guide" and "User
Review" sections.

This script will let you specify the range of title ID numbers to search, the
keyword(s) you want to search for, and other options such as delay between
page requests and output log file.

Important!
==============

The TOS of the IMDB specifically forbids accessing their data through automated
methods such as this, and it's possible you can have your IP banned if you
abuse their service.

Understand that, if you use this script you're technically in violation of
IMDB's rules and regulations, and that there is no telling what (if anything)
will happen to you.

If you decide to use this software anyway, make use of the SCAN_DELAY option
in the configuration to wait a reasonable time before requesting a new page.

Functions
==============

Running imdb_scrape with no arguments will start it in default mode, which
searches IMDB based on the parameters setup in either the external
configuration file or the script's internal configuration section.

In addition, there are a few other functions which can be accessed by typing
the function name after the script. They are:

#### download
This function will read a previously created file of matching movies, and
download their pages. This function should be used if you intend on doing any
local processing of the files, removing the need to repeatedly pull the files
down from IMDB.

#### process
This function attempts to trim and condense the local files to make them more
manageable. It will remove a lot of the HTML from the downloaded files, and
merge them together into one file per movie which contains just (more or less)
the user-content.

#### clean
Remove the downloaded files. This will not remove the files created with the
"process" function, only the raw HTML which was saved with "download".

#### help
A quick intro to the script and listing of available functions.

Requirements
==============

This script requires a number of common Unix tools:

* sed
* awk
* wget
* grep

Every Unix/Linux distribution should include these tools, or at the very least,
have a way for you to easily install them.

It has been tested on Ubuntu, Mint, Slackware, and OpenWRT with no issues.

License
==============

This program is free software; you can redistribute it and/or modify it under
the terms of the BSD 3-Clause license.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

For details, see the "LICENSE" file.


ChangeLog
12/15/13:
	Warn on missing files during process
	Improve error handling on merge
	Improve processed file trimming
	Add title/ID to processed file
	Remove merge tmp files	

12/14/13: Released as version 1.3
	Push more code into its own functions
	Fix issue with movie titles that have commas
	Added "Download" mode
	Added "Process" mode
	Added help screen
	Updated docs

12/12/13: Released as version 1.2
	Exit after user-defined number of 404 pages.
	Optionally check if pages are actually from IMDB

12/04/13: Released as version 1.1
	Gracefully handle 404 pages
	Skip delay on last fetch
	Improved boilerplate
	Fix rm statement
	Fix comments

12/02/13:
	Add support for external config file
	Example config file added

12/02/13: Released as version 1.0
	First public release, pretty much works.


imdb_scrape.sh
#!/bin/bash
#
# imdb_scrape: Search user-submitted IMDB data for keywords.
VER="1.3"
#
# This software is released under the BSD 3-Clause license. See the included
# file "LICENSE" for more information.
#
#-------------------------User Configuration Section---------------------------#

# Starting title
TT_START=87332

# Ending title
TT_END=87340

# Keyword
KEYWORD='spirit'

# Output file
LOG_FILE="results.txt"

# Delay in seconds
SCAN_DELAY=15

# Delete files after processing
KEEP_FILES=0

# Number of 404's to accept before exiting
ERR_LIMIT=5

# Make sure page is valid IMDB entry before processing
IMDB_CHECK=0

# Configuration file, overrides above settings
CONFIG_FILE="scrape.conf"

#------------------------No need to edit past this line------------------------#

# Source external file to override local config
. $CONFIG_FILE

# URL parts
URL_BASE="http://www.imdb.com/title/tt"
PAGE1="/parentalguide"
PAGE2="/reviews"

# Temp directory
TMP_DIR="/tmp"

ErrorMsg ()
{
# ErrorMsg  
# Displays either a minor (warning) or critical error, exiting on an critical.
# If message starts with "n", a newline is printed before message.
[[ $(expr substr "$2" 1 1) == "n" ]] && echo
if [ "$1" == "ERR" ]; then
	# This is a critical error, game over.
	echo "  ERROR: ${2#n}"
	exit 1
elif [ "$1" == "WRN" ]; then
	# This is only a warning, script continues but may not work fully.
	echo "  WARNING: ${2#n}"
fi
}

VerifyCmd ()
{
# VerifyCmd  
# Checks to see if given command exists, optional output.
[ $1 -eq 1 ] && echo -n "Checking for $2: "
if which $2 > /dev/null 2>&1; then
	[ $1 -eq 1 ] && echo "OK"
	return 0
else
	[ $1 -eq 1 ] && echo "FAILED"
	return 1
fi
}

CheckSys ()
{
# Check if tools are installed
VerifyCmd $1 grep
[ $? -eq 1 ] && ErrorMsg ERR "grep not found! Install to continue."
VerifyCmd $1 sed
[ $? -eq 1 ] && ErrorMsg ERR "sed not found! Install to continue."
VerifyCmd $1 awk
[ $? -eq 1 ] && ErrorMsg ERR "awk not found! Install to continue."
VerifyCmd $1 wget
[ $? -eq 1 ] && ErrorMsg ERR "wget not found! Install to continue."
}

StartUp ()
{
# Sys check, boilerplate
clear
CheckSys 0
echo "imdb_scrape $VER"
echo "----------------"
}

DownloadPage ()
{
# Download the pages for TT_CUR
if wget -q $URL_BASE$TT_CUR$PAGE1 -O $TMP_DIR/PARENT$TT_CUR ; then
	wget -q $URL_BASE$TT_CUR$PAGE2 -O $TMP_DIR/REVIEW$TT_CUR
else
	echo "404 Page"
	((ERR_COUNT++))
	if [ $ERR_COUNT -eq $ERR_LIMIT ] ; then
		ErrorMsg ERR "Too many 404 pages, exiting."
		exit
	fi
		continue
fi 
}

GetTitle ()
{
# Return current movie title
TITLE=`grep "<title>" $TMP_DIR/PARENT$TT_CUR | awk -F"[><]" '{ print $3 }' | sed 's/.\{16\}$//'`
}

VerifyIMDB ()
{
# Verify the page we are looking at is actually from IMDB
grep -iq "imdb" $TMP_DIR/PARENT$TT_CUR || ErrorMsg ERR "nNot an IMDB page!"
}

#---------------------------Execution starts here------------------------------#
case $1 in
'check')
echo "Checking system..."
CheckSys 1
;;
'clean')
echo "Removing old temp files..."
rm -f $TMP_DIR/PARENT*
rm -f $TMP_DIR/REVIEW*
rm -f $TMP_DIR/IMDB_TMP*
;;
'help')
clear
echo "imdb_scrape $VER"
echo "----------------"
echo "This script can be used to search the user-submitted data on IMDB.com for"
echo "specific keywords and phrases. The name of each movie that matches the"
echo "search terms will be written to the log file, along with the movie's"
echo "IMDB ID number."
echo
echo "Optionally, the script can save the matching pages, trimming and condensing"
echo "them down to a single file for each movie."
echo
echo "WARNING!"
echo "Use of this script is in direct violation of the IMDB Terms of Use."
echo
echo "The available arguments as of version $VER are as follows:"
echo "download   - Read the log file and download corresponding pages"
echo "process    - Trim down and merge pages into one file per movie"
echo "clean      - Remove downloaded files"
echo "help       - What you are reading now"
;;
'process')
StartUp
echo "Reading from: $LOG_FILE" 
echo "Processing files from: $TMP_DIR"
echo
# Loop through file
while read LINE
do
	# Read current title #
	TT_CUR=`echo $LINE | awk -F: '{print $1}'`
	echo "Processing $LINE"

	# Remove old temp files
	rm -f $TMP_DIR/IMDB_TMP*
	
	# Check for parent's guide and process
	if [ -s $TMP_DIR/PARENT$TT_CUR ] ; then
		# Remove all lines before first parent's guide category
		sed -i '1,/Sex &amp; Nudity/d' $TMP_DIR/PARENT$TT_CUR

		# Remove lines after "Report Problem", put into new file
		sed -n '/Report a problem/q;p' $TMP_DIR/PARENT$TT_CUR > $TMP_DIR/IMDB_TMP1
	else
		ErrorMsg WRN "Parents Guide not found."
	fi

	# Check for review and process
	if [ -s $TMP_DIR/REVIEW$TT_CUR ] ; then
		# Remove all lines before first horizonal bar
		sed -i '1,/<hr size="1" noshade="1">/d' $TMP_DIR/REVIEW$TT_CUR

		# Remove lines after second bar, put into new file
		sed -n '/<hr size="1" noshade="1">/q;p' $TMP_DIR/REVIEW$TT_CUR > $TMP_DIR/IMDB_TMP2
	else
		ErrorMsg WRN "Reviews not found."
	fi

	# Create final file, first line is movie title/ID
	echo $LINE > $TMP_DIR/TITLE_$TT_CUR

	# Merge together
	cat $TMP_DIR/IMDB_TMP* >> $TMP_DIR/TITLE_$TT_CUR 2>/dev/null || ErrorMsg ERR "Files not found!"
done < $LOG_FILE
echo "Complete"
;;
'download')
StartUp
echo "Reading from: $LOG_FILE" 
echo "Downloading pages to: $TMP_DIR"
echo
# Loop through file
while read LINE
do
	# Read current title #
	TT_CUR=`echo $LINE | awk -F: '{print $1}'`

	# Display entry
	echo $LINE

	# Download pages for current title
	DownloadPage
	
	# Delay next fetch	
	sleep $SCAN_DELAY
done < $LOG_FILE
;;
*)
StartUp
echo "Term: $KEYWORD"
echo "Titles: $TT_START through $TT_END"
echo
# Loop through range
for((TT_CUR=$TT_START;TT_CUR<=$TT_END;++TT_CUR)) do

	# Download pages for current title
	echo -n "Fetching title $TT_CUR: "
	DownloadPage

	# See if this is an IMDB page
	[ $IMDB_CHECK -eq 1 ] && VerifyIMDB

	# Get and display movie title
	GetTitle
	echo -n $TITLE	

	# Run through grep
	if grep -iq $KEYWORD $TMP_DIR/*$TT_CUR ; then
		# Match found!
		echo " - HIT!"
		# Log title
		echo "$TT_CUR: $TITLE" >> $LOG_FILE
	else
		echo
	fi
	
	# Delete files
	[ $KEEP_FILES -eq 0 ] && rm -f $TMP_DIR/*$TT_CUR

	# Delay next fetch	
	[ $TT_CUR -ne $TT_END ] && sleep $SCAN_DELAY
done
;;
esac
#EOF
